src/planet-a/SubdivSkeleton.h

```diff
@@ -0,0 +1,541 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/id_registry.h>
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+#include <optional>
+#include <limits>
+#include <unordered_map>
+#include <vector>
+
+namespace planeta
+{
+
+template<typename T>
+using ArrayView_t = Corrade::Containers::ArrayView<T>;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A multitree directed acyclic graph of reusable IDs where new IDs can
+ *        be created from two other parent IDs.
+ */
+template<typename ID_T>
+class SubdivIdTree : private osp::IdRegistry<ID_T>
+{
+
+    using base_t = osp::IdRegistry<ID_T>;
+    using id_int_t = osp::underlying_int_type_t<ID_T>;
+
+    static_assert(std::is_integral_v<id_int_t> && sizeof(ID_T) <= 4,
+                  "ID_T must be an integral type, 4 bytes or less in size");
+
+    using combination_t = uint64_t;
```

Comment:
Is this just because `uint64_t` is the max size of two 32-bit indices? That makes sense, but it feels a little weird to use it explicitly here when the class is so heavily templatized, but I suppose this is the most reasonable way to do it. Might want to write a quick comment explaining this; even though there's a good clue in the `static_assert()` just above it's still not blaringly obvious and if it weren't for the hint above I would not have figured it out.

Reply:
I didn't want to spend time writing a "get an integer type 2x larger than the given integer type" template when this class is going to be used like 2 times lol. Right now this isn't run anywhere, so I'll leave it for a next PR once it's actually used.